File: plugins/repository/src/main/java/no/lillehaug/landingsopplysninger/repository/jdbc/JDBC.java
package no.lillehaug.landingsopplysninger.repository.jdbc;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.SQLException;
import java.util.function.Function;

public class JDBC {

    /*public static <T> T readOnly(DataSource ds, Function<Connection, T> action) throws SQLException {
        Connection c = ds.getConnection();
        boolean readOnly = c.isReadOnly();
        c.setReadOnly(true);
        try(c) {
            return action.apply(c);
        } finally {
            c.setReadOnly(readOnly);
        }
    }

    public static <T> T transactional(DataSource ds, Function<Connection, T> action) throws SQLException {
        Connection c = ds.getConnection();
        boolean autoCommit = c.getAutoCommit();
        c.setAutoCommit(false);
        try(c) {
            return action.apply(c);
        } catch (Throwable e) {
            c.rollback();
            throw e;
        } finally {
            c.setAutoCommit(autoCommit);
        }
    }*/
}

File: plugins/repository/src/main/java/no/lillehaug/landingsopplysninger/repository/jdbc/JDBC.kt
package no.lillehaug.landingsopplysninger.repository.jdbc

import no.lillehaug.landingsopplysninger.library.TryWR.Companion.trywr
import org.slf4j.LoggerFactory
import java.sql.Connection
import javax.sql.DataSource

class JDBC {
    companion object {
        val log = LoggerFactory.getLogger(JDBC::class.java)

        @JvmStatic
        fun <T> readOnly(ds: DataSource, action: (c: Connection) -> T): T {
            val connection = ds.connection
            val isReadOnly = connection.isReadOnly
            connection.isReadOnly = true
            return trywr(connection) {
                val result = action(connection)
                connection.isReadOnly = isReadOnly
                result
            }
        }

        @JvmStatic
        fun <T> transactional(ds: DataSource, action: (c: Connection) -> T): T {
            val connection = ds.connection
            connection.autoCommit = false
            return trywr(connection) {
                try {
                    val result = action(connection)
                    connection.commit()
                    result
                } catch(e: Exception) {
                    log.error("Exception in transaction", e)
                    connection.rollback()
                    throw e
                }
            }
        }


    }
}
